### PR TITLE
SCIM authentication

### PR DESCRIPTION
**Pull request by neongreen**
The first commit here lets us use new hscim and token-based authentication. However, we also need a way to create and store SCIM tokens for teams.

I would propose adding a new table:

```
CREATE TABLE if not exists team_provisioning
    ( team          uuid
    , token         text
    , idp           uuid  -- optional
    , PRIMARY KEY (team, token)
    );
```

This way we can:

  1. Quickly check whether a token is valid.

  2. Have teams without provisioning support (good for reducing attack surface). An API endpoint would be added to generate a token for a team or to clear all existing tokens.

  3. Allow provisioning without an IdP (Wire Red). For now we'll just throw an error if there's a token but no IdP attached to it, because we need the IdP to create a user in Spar.

### CODE REVIEW

**Comment by neongreen** [comment]
> if there is more than one token, it should be possible to pick one for deletion and keep the others.

Well, sure, or that.

File: services/spar/src/Spar/SCIM.hs [hunk 1]
```diff
@@ -306,23 +289,37 @@ instance SCIM.UserDB (ReaderT TeamId Spar) where
...
+  type AuthData Spar = ScimToken
+  type AuthInfo Spar = TeamId
+
   -- TODO
-  authCheck Nothing =
-    pure Unauthorized
-  authCheck (Just (SCIM.SCIMAuthData uuid _pass)) =
-    pure (Authorized (SCIM.Admin uuid))
+  authCheck _ = SCIM.throwSCIM $
+      SCIM.unauthorized "authorization not implemented"
```
[the remark this answers is not in the document]
Comment (neongreen) [reply services/spar/src/Spar/SCIM.hs]: I will test that bad tokens are rejected when we have a table (or some other way to test tokens). Currently there's an "acceptable token" in SCIM test settings but it's actually never used and never accessed. (That's why this PR is WIP!)

**Comment by neongreen** [comment]
* [ ] Also, if we decide to go with tokens, let's make sure they are HMAC-ed.

**Comment by fisx** [comment]
> I'll avoid doing anything with this PR until then.

donno, i thought the decision to continue with HTTP basic auth stands, and the ongoing discussion is about how to improve on that once we are live with that.  so in my (unauthoritative) opinion you can easily finish that with what we know now.

**Comment by neongreen** [comment]
Unresolved questions:

1. What library to use for HMAC? What are we already using, if anything?
2. What endpoints should we create for token creation and revocation? Naming things is hard.
3. Should we have (for now) a limit of one token per team? If not, what should be the limit?

**Comment by fisx** [comment]
> * What library to use for HMAC? What are we already using, if anything?

we already depend on cryptonite, which [has this](http://hackage.haskell.org/package/cryptonite-0.25/docs/Crypto-MAC-HMAC.html).

> * What endpoints should we create for token creation and revocation? Naming things is hard.

- `POST /scim/auth-token`
- `DELETE /scim/auth-token`
- `DELETE /scim/auth-token/:token`

(are the routing tables coming from hscim located under an `/scim/` prefix?)

> * Should we have (for now) a limit of one token per team? If not, what should be the limit?

12!  (for no strong reason)

**Comment by neongreen** [comment]
> (are the routing tables coming from hscim located under an /scim/ prefix?)

No, they are prefix-less. We add the `/scim/` prefix in Spar.

**Comment by neongreen** [comment]
I'm very mildly uneasy about using `/scim/` for tokens that may one day be used for provisioning via things other than SCIM (yeah the token type is `ScimToken` but it's much easier to rename in the future than an endpoint). However, emphasis on "very mildly".

**Comment by fisx** [comment]
> I'm very mildly uneasy about using `/scim/` for tokens that may one day be used for provisioning via things other than SCIM (yeah the token type is `ScimToken` but it's much easier to rename in the future than an endpoint).

if that happens, we could have two end-points serving the same handler.  i don't insist on `/scim/`, but it's nice to have the token under the same prefix as the rest of the protocol, and i can't think of any other name.

File: services/galley/src/Galley/Intra/Spar.hs [hunk 2]
```diff
@@ -0,0 +1,24 @@
+{-# LANGUAGE OverloadedStrings #-}
+
+module Galley.Intra.Spar
```
Comment (fisx) [inline services/galley/src/Galley/Intra/Spar.hs]: Shouldn't this be a separate PR?

Comment (neongreen) [reply services/galley/src/Galley/Intra/Spar.hs]: I can make it a separate PR, yeah.

Comment (neongreen) [reply services/galley/src/Galley/Intra/Spar.hs]: I have expelled that code into #519.

File: services/spar/src/Spar/SCIM.hs [hunk 3]
```diff
@@ -303,23 +318,111 @@ instance SCIM.UserDB (ReaderT TeamId Spar) where
...
+  type AuthData Spar = ScimToken
+  type AuthInfo Spar = (TeamId, Maybe IdPId)
+
   authCheck Nothing =
-    pure Unauthorized
-  authCheck (Just (SCIM.SCIMAuthData uuid _pass)) =
-    pure (Authorized (SCIM.Admin uuid))
+      SCIM.throwSCIM (SCIM.unauthorized "Token not provided")
+  authCheck (Just token) =
+      maybe (SCIM.throwSCIM (SCIM.unauthorized "Invalid token")) pure =<<
```
Comment (gromakovsky) [inline services/spar/src/Spar/SCIM.hs]: Why does the new code use 4 spaces indentation while the old code uses 2 spaces?

Comment (fisx) [reply services/spar/src/Spar/SCIM.hs]: because it can?  :-)

we are thinking about using something like [brittany](http://hackage.haskell.org/package/brittany) to resolve all these discussions once and for all, but until that's ready for our code base, we try not to spend too much time on being too strict about it.

Comment (neongreen) [reply services/spar/src/Spar/SCIM.hs]: I usually do the same indentation as the file already has, but Spar has a mix of indentation styles so sometimes I just go with what my editor suggests.

**Comment by neongreen** [comment]
NB: this shouldn't be merged until PRs to cailleach and hegemony are done.

File: services/spar/test-integration/Test/Spar/SCIMSpec.hs [hunk 4]
```diff
@@ -115,12 +116,57 @@ randomUser = do
...
+            -- that it fails
+            listUsers_ Nothing Nothing (env ^. teSpar)
+                !!! const 401 === statusCode
+            -- Try to execute the same SCIM operation with the generated
+            -- token; it should succeed now
+            listUsers_ (Just token) Nothing (env ^. teSpar)
+                !!! const 200 === statusCode
+            -- Cleanup
+            deleteToken token
+        it "respects the token limit (2 for integration tests)" $ do
```
Comment (fisx) [inline services/spar/test-integration/Test/Spar/SCIMSpec.hs]: If it is only ok to have 1 token, but not 2, then the limit should be `1`, no?

File: services/spar/test-integration/Test/Spar/SCIMSpec.hs [hunk 5]
```diff
@@ -38,7 +39,7 @@ import qualified Data.Aeson as Aeson
 spec :: SpecWith TestEnv
 spec = do
     specUsers
-
+    specTokens
```
Comment (fisx) [inline services/spar/test-integration/Test/Spar/SCIMSpec.hs]: missing tests (i think; correct me if i am wrong):

- if idp is not provided, the create-token request fails (or is it the SAML login request later on that fails?)
- come to think of it, test that SCIM and saml work together and SCIM-created users can login.  give this task back to me if you like, I've recently written a lot of test code that can be re-used for this.
- `Data.getScimTokens` is not tested here.  I guess you may as well add an end-point for that now in `APIScimToken`

(I personally prefer to test the database interface separately, but I don't want to insist.)

File: services/spar/src/Spar/SCIM.hs [hunk 6]
```diff
@@ -271,11 +272,17 @@ instance SCIM.UserDB Spar where
...
-    let uref = SAML.UserRef
-          (SAML.Issuer $ SAML.unsafeParseURI ("https://unknown.issuer.org"))
-          (SAML.opaqueNameID extId)
+    -- to extract that later
+    -- TODO: when the issuer is deleted, the token still remains, so we can
+    -- fail here
+    -- TODO: throw SCIM errors here
+    issuer <- case mbIdp of
+        Nothing -> error "No IdP configured for the provisioning token"
+        -- TODO: should 'wrapMonadClient' or 'wrapMonadClientWithEnv' be used here?
```
Comment (fisx) [inline services/spar/src/Spar/SCIM.hs]: `wrapMonadClientWithEnv` gives you more context than `MonadClient`, so only use it if you need that context.  Currently `Env` carries the current system time and two SAML-specific TTL-values.

File: services/spar/src/Spar/SCIM.hs [hunk 7]
```diff
@@ -271,11 +272,17 @@ instance SCIM.UserDB Spar where
...
     -- TODO: Assume that externalID is the subjectID, let's figure out how
-    -- to extract that later. The question of "who's the issuer?" is a very
-    -- tricky one and we don't have a decision here yet.
-    let uref = SAML.UserRef
-          (SAML.Issuer $ SAML.unsafeParseURI ("https://unknown.issuer.org"))
-          (SAML.opaqueNameID extId)
+    -- to extract that later
+    -- TODO: when the issuer is deleted, the token still remains, so we can
+    -- fail here
+    -- TODO: throw SCIM errors here
```
Comment (fisx) [inline services/spar/src/Spar/SCIM.hs]: please resolve this TODO and handle the errors.

File: services/spar/src/Spar/SCIM.hs [hunk 8]
```diff
@@ -271,11 +272,17 @@ instance SCIM.UserDB Spar where
...
     buid <- Id <$> liftIO UUID.nextRandom
     -- TODO: Assume that externalID is the subjectID, let's figure out how
-    -- to extract that later. The question of "who's the issuer?" is a very
-    -- tricky one and we don't have a decision here yet.
-    let uref = SAML.UserRef
-          (SAML.Issuer $ SAML.unsafeParseURI ("https://unknown.issuer.org"))
-          (SAML.opaqueNameID extId)
+    -- to extract that later
+    -- TODO: when the issuer is deleted, the token still remains, so we can
+    -- fail here
```
Comment (fisx) [inline services/spar/src/Spar/SCIM.hs]: this should be fine, no?  just throw an appropriate error.  please resolve this TODO!

File: services/spar/test-integration/Util.hs [hunk 9]
```diff
@@ -166,7 +166,14 @@ mkEnv _teTstOpts _teOpts = do
...
 
-  let _teScimToken = ScimToken "scim-test-token"
+  let _teScimToken = ScimToken $
+          "scim-test-token/" <> "team=" <> idToText _teTeamId
+  runClient _teCql $
+      Data.insertScimToken
+          _teTeamId
+          _teScimToken
+          (Just (_teIdP ^. idpId))
+          "_teScimToken test token"
```
Comment (fisx) [inline services/spar/test-integration/Util.hs]: Btw, can you add a haddock comment about what the description is (or will be) used for?

File: services/spar/test/Test/Spar/APISpec.hs [hunk 10]
```diff
@@ -17,11 +17,5 @@ import Test.Hspec
...
-  validateEveryToJSON (Proxy :: Proxy API.APIMeta)
-  validateEveryToJSON (Proxy :: Proxy API.APIAuthReqPrecheck)
-  validateEveryToJSON (Proxy :: Proxy API.APIAuthReq)
-  validateEveryToJSON (Proxy :: Proxy API.APIAuthResp)
-  validateEveryToJSON (Proxy :: Proxy API.IdpGet)
-  validateEveryToJSON (Proxy :: Proxy API.IdpGetAll)
-  validateEveryToJSON (Proxy :: Proxy API.IdpCreate)
-  validateEveryToJSON (Proxy :: Proxy API.IdpDelete)
+  -- Note: SCIM types are not validated because their content-type is 'SCIM'.
+  validateEveryToJSON (Proxy @API.OutsideWorldAPI)
```
Comment (fisx) [inline services/spar/test/Test/Spar/APISpec.hs]: :+1: thanks!  :-)

File: services/spar/src/Spar/SCIM.hs [hunk 11]
```diff
@@ -399,6 +399,10 @@ createScimToken :: Maybe UserId -> CreateScimToken -> Spar ScimToken
 createScimToken zusr CreateScimToken{..} = do
     let descr = createScimTokenDescription
     teamid <- getZUsrOwnedTeam zusr
+    tokenNumber <- fmap length $ wrapMonadClient $ Data.getScimTokens teamid
+    maxTokens <- asks (maxScimTokens . sparCtxOpts)
+    unless (tokenNumber < maxTokens) $
```
Comment (fisx) [inline services/spar/src/Spar/SCIM.hs]: Ah!  This should really be `<=`, I think.  (See my comment on the related test case.)

File: services/spar/test-integration/Test/Spar/SCIMSpec.hs (hunk 4 above)
Comment (neongreen) [reply services/spar/test-integration/Test/Spar/SCIMSpec.hs]: The limit is "maxScimTokens is allowed, maxScimTokens+1 is not allowed". One token is already present because we created it in TeamEnv. I'll add a comment clarifying that,

File: services/spar/src/Spar/SCIM.hs (hunk 8 above)
Comment (neongreen) [reply services/spar/src/Spar/SCIM.hs]: NB: Ideally we should delete the token when the issuer is deleted, but maybe that's hard to pull off and that's not done in this PR.

File: services/spar/src/Spar/SCIM.hs (hunk 11 above)
Comment (neongreen) [reply services/spar/src/Spar/SCIM.hs]: I think it should be `<`. If we already have `maxTokens` tokens, we can't add a new one. So we should fail unless there's space for at least one more token.

Comment (fisx) [reply services/spar/src/Spar/SCIM.hs]: right.  my mistake, sorry.

File: services/spar/src/Spar/SCIM.hs (hunk 8 above)
Comment (fisx) [reply services/spar/src/Spar/SCIM.hs]: > NB: Ideally we should delete the token when the issuer is deleted, but maybe that's hard to pull off and that's not done in this PR.

what do you mean by `Issuer`?  the idp?  then what about idp-less tokens?

i thought that tokens should last as long as idp (if there is one) or team (if there isn't), or if they are deleted manually.

1. with idp: just add a line or two to the `Data.deleteIdP` (or whatever it's called).
2. without idp: oh, right.  i asked you to move that to #519.  i guess this TODO should also be resolved there, then.  :-)
3. manually: already done.

File: services/spar/test-integration/Test/Spar/SCIMSpec.hs (hunk 5 above)
Comment (neongreen) [reply services/spar/test-integration/Test/Spar/SCIMSpec.hs]: @fisx 

> give this task back to me if you like, I've recently written a lot of test code that can be re-used for this.

Would you like to do it in a new PR, or in this one?

Comment (neongreen) [reply services/spar/test-integration/Test/Spar/SCIMSpec.hs]: > if idp is not provided, the create-token request fails (or is it the SAML login request later on that fails?)

You don't provide the IdP when creating the token; the backend looks at the list of IdPs for your team and fails unless there's exactly one. I can add a test like "token can't be created for an IdP-less team", though.

Comment (fisx) [reply services/spar/test-integration/Test/Spar/SCIMSpec.hs]: > @fisx
> 
> > give this task back to me if you like, I've recently written a lot of test code that can be re-used for this.
> 
> Would you like to do it in a new PR, or in this one?

ok, fine, i'll do it somewhere else.  :-)  consider it solved!

File: services/spar/src/Spar/SCIM.hs (hunk 8 above)
Comment (neongreen) [reply services/spar/src/Spar/SCIM.hs]: > with idp: just add a line or two to the Data.deleteIdP (or whatever it's called).

Done.

> without idp

Yeah, will be done in #519.

File: services/spar/src/Spar/SCIM.hs (hunk 7 above)
Comment (neongreen) [reply services/spar/src/Spar/SCIM.hs]: Done. (Not sure what handling them would entail, though. It's error500.)

File: services/spar/test-integration/Test/Spar/SCIMSpec.hs (hunk 5 above)
Comment (neongreen) [reply services/spar/test-integration/Test/Spar/SCIMSpec.hs]: > Data.getScimTokens is not tested here. I guess you may as well add an end-point for that now in APIScimToken

I'll do that and test it.

> I personally prefer to test the database interface separately.

Can you very briefly describe what kind of database tests you'd like to see? Like, in pseudocode.

File: services/spar/src/Spar/API/Swagger.hs [hunk 12]
```diff
@@ -156,3 +149,26 @@ instance ToParamSchema BindCookie where
...
   declareNamedSchema _ = declareNamedSchema (Proxy @String)
+
+instance ToParamSchema ScimToken where
+  toParamSchema _ = toParamSchema (Proxy @Text)
+
+instance ToSchema ScimToken where
+  declareNamedSchema _ = declareNamedSchema (Proxy @Text)
+    & mapped . schema . description ?~ "Authentication token"
+
+instance ToSchema ScimTokenInfo where
```
Comment (neongreen) [inline services/spar/src/Spar/API/Swagger.hs]: Should I create Spar.SCIM.Swagger and move these there?

Comment (neongreen) [reply services/spar/src/Spar/API/Swagger.hs]: Yeah, I think I should.

Comment (fisx) [reply services/spar/src/Spar/API/Swagger.hs]: I have no idea which would be better.  It's nice to have the docs in one place, but it's also nice to have the docs in the place where the code is.  Your call!

File: services/spar/src/Spar/Data.hs [hunk 13]
```diff
@@ -373,24 +379,24 @@ lookupScimToken
...
+-- | List all tokens associated with a team, in the order of their creation.
 getScimTokens
   :: (HasCallStack, MonadClient m)
   => TeamId -> m [ScimTokenInfo]
 getScimTokens team = do
     -- We don't need pagination here because the limit should be pretty low
     -- (e.g. 16). If the limit grows, we might have to introduce pagination.
     rows <- retry x1 . query sel $ params Quorum (Identity team)
-    pure $ map (\(_, stiTeam, stiId, stiIdP, stiDescr) -> ScimTokenInfo{..}) rows
+    pure $ sortOn stiCreatedAt $ map fromScimTokenRow rows
```
Comment (fisx) [inline services/spar/src/Spar/Data.hs]: Can't cassandra do the sorting more efficiently here?

File: services/spar/src/Spar/SCIM.hs [hunk 14]
```diff
@@ -456,12 +456,14 @@ createScimToken zusr CreateScimToken{..} = do
             -- it makes sense semantically
             token <- ScimToken . cs . ES.encode <$> liftIO (randBytes 32)
             tokenid <- randomId
+            now <- liftIO getCurrentTime
```
Comment (fisx) [inline services/spar/src/Spar/SCIM.hs]: The `Spar` monad has `getNow` from saml2-web-sso, but I'm not sure you want to use that in this module.  There is something nice about not having SCIM depend on SAML.

So, feel free to ignore this comment.

File: services/spar/src/Spar/Data.hs (hunk 13 above)
Comment (neongreen) [reply services/spar/src/Spar/Data.hs]: Not without making creation time a clustering key. Which actually might not be a bad idea 🤔 @tiago-loureiro what do you think?

```diff
    void $ schema' [r|
        CREATE TABLE if not exists team_provisioning_by_team
            ( token_        text
            , team          uuid
            , id            uuid
            , created_at    timestamp
            , idp           uuid         -- optional
            , descr         text
-           , PRIMARY KEY (team, id)
+           , PRIMARY KEY (team, id, created_at)
            ) with compaction = {'class': 'LeveledCompactionStrategy'}
+             and clustering order by (created_at asc);
    |]
```

Comment (tiago-loureiro) [reply services/spar/src/Spar/Data.hs]: > Which actually might not be a bad idea 

No objections here. If u want to keep a limit, bear in mind how u want to limit those cookies (do u delete the last one, etc.)? I.e., adding a created_at / expires_at might be useful indeed, I'd just try to optimise it so that deleting/overwriting cookies is straightforward. (And yeah, 16 such cookies sounds like plenty to me).

Comment (neongreen) [reply services/spar/src/Spar/Data.hs]: > If u want to keep a limit, bear in mind how u want to limit those cookies (do u delete the last one, etc.)?

These are not cookies and they are not supposed to ever disappear implicitly (or be overwritten). If you've reached your limit, we're just not going to let you add a new one.
